NEW: Feature Stain Normalisation Ruifrok & Reinhard

simongraham opened the pull request:
Added stain normalisation code based on the [StainTools](https://github.com/Peter554/StainTools) library. Methods implemented:

- Reinhard colour normalisation
- Ruifrok & Johnston stain normalisation

Files added/modified:

- `tiatoolbox/tools/stainnorm.py`
- `tiatoolbox/utils/stainnorm.py`
- `tests/test_tiatoolbox.py`
- `tiatoolbox/cli.py`
- `docs/usage.rst`

Requiring attention -> please check the tests. This is done based on storing normalised images in the repo and comparing the results to those images.

Not yet implemented from StainTools:

- Tissue mask method
- Brightness standardiser

Need to discuss whether to add these.

shaneahmed commented on tests/test_tiatoolbox.py:
```diff
@@ -6,8 +6,11 @@
 from tiatoolbox.dataloader.save_tiles import save_tiles
 from tiatoolbox.dataloader import wsireader
+from tiatoolbox.tools.stainnorm import reinhard_colour_normaliser
+from tiatoolbox.tools.stainnorm import stain_normaliser
```
Please use 'from tiatoolbox.tools.stainnorm import reinhard_colour_normaliser, stain_normaliser` instead of importing in two lines.

simongraham replied:
Fixed in [this commit](https://github.com/TIA-Lab/tiatoolbox/pull/35/commits/46b2c26c5dc5862f1e4106f2a5347ddb54a59d46)

shaneahmed commented on tiatoolbox/cli.py:
```diff
@@ -66,9 +66,18 @@ def main():
 )
 @click.option(
-    "--verbose", type=bool, default=True, help="Print output, default=True",
+    "--workers",
```
We removed multiproc from develop branch, best to merge latest develop branch into this branch to avoid these issues.

simongraham replied:
Fixed in [this commit](https://github.com/TIA-Lab/tiatoolbox/pull/35/commits/47deccd6060278762872c1930d14707625aa2b2d)

shaneahmed commented on tiatoolbox/tools/stainnorm/reinhard_colour_normaliser.py:
```diff
@@ -0,0 +1,114 @@
+        """Convert from RGB uint8 to LAB and split into channels.
+
+        Args:
+            img (RGB uint8): Input image
```
ndarray instead of RGB.

simongraham replied:
Fixed in [this commit](https://github.com/TIA-Lab/tiatoolbox/pull/35/commits/38af1ea1723fb76a8c553e6468a60ddc6c75c999)

shaneahmed commented on tests/test_tiatoolbox.py:
```diff
@@ -408,3 +418,41 @@ def test_command_line_save_tiles(_response_ndpi, _response_svs):
+
+def test_reinhard_normalise():
+    """Test Reinhard colour normalisation."""
+    source_img = imread("tiatoolbox/tools/stainnorm/samples/source.png")
```
Do we need to include these images? Can we upload them on tia webpage and get access from there?

simongraham replied:
Yes this is a good idea and was something I wanted to discuss too. Where exactly can we upload them?

shaneahmed replied:
You can upload [here](https://warwick.ac.uk/fac/sci/dcs/research/tia/tiatoolbox/files)

simongraham replied:
Addressed in [this commit](https://github.com/TIA-Lab/tiatoolbox/pull/35/commits/3cfb8fe5f2b77998950484dfed6bd504af4670dd).

shaneahmed commented on tiatoolbox/cli.py:
```diff
@@ -66,7 +66,10 @@ def main():
     "the meta information, default=show",
```
Did you add a cli function for stain normalisation? This needs to be added and also need to write a test for it.

simongraham replied:
No, I didn't add that. I Will do it.

shaneahmed commented on tests/test_tiatoolbox.py:
```diff
@@ -408,3 +418,41 @@ def test_command_line_save_tiles(_response_ndpi, _response_svs):
+
+    reinhard_img = imread("tiatoolbox/tools/stainnorm/samples/reinhard.png")
+
+    norm = reinhard_colour_normaliser.ReinhardColourNormaliser()
```
Can we define the classes in a single file? e.g., we can have stainnorm.ReinhardColourNormaliser() and stainnorm.Ruifork()?

simongraham replied:
The way of doing this that I can think of would be to create a script `tiatoolbox/tools/stainnorm.py` and then defining each stain normalisation class within there. Then we can have a stain_utils directory for the utils. What do you think?

shaneahmed replied:
Yes sounds good to me.

simongraham replied:
Modified in [this commit](https://github.com/TIA-Lab/tiatoolbox/pull/35/commits/ad1209a509e700c326ff847fa61437b264294582).

shaneahmed commented on tiatoolbox/tools/stainnorm/reinhard_colour_normaliser.py:
```diff
@@ -0,0 +1,114 @@
+    """Normalize a patch colour to the target image using the method of:
+
+    Reinhard et al. "Color transfer between images." Computer graphics & applications.
+
```
Need to define attributes for this class.

simongraham replied:
Can you define what you mean by attributes here? Class attributes or instance attributes?

shaneahmed replied:
Please consider the example at this [link](https://github.com/TIA-Lab/tiatoolbox/blob/a826f658dd02671631d83cfc42676b2858548fb5/tiatoolbox/dataloader/wsireader.py#L35)

simongraham replied:
This isn't really clear by looking at that example. Are you referring to the variables defined as arguments in the __init__ or the instance variables defined in __init__ itself? Here, I didn't include attributes because __init__ only takes self as an argument.

shaneahmed replied:
self.target_means = means
self.target_stds = stds

shaneahmed commented on tiatoolbox/tools/stainnorm/stain_normaliser.py:
```diff
@@ -0,0 +1,80 @@
+    """Stain normalisation class
+
+    Attributes:
+        method (string): stain normalisation method to use
```
attributes are not complete

simongraham replied:
Can you define what you mean by attributes here? Class attributes or instance attributes?

shaneahmed replied:
Please consider the example at this [link](https://github.com/TIA-Lab/tiatoolbox/blob/a826f658dd02671631d83cfc42676b2858548fb5/tiatoolbox/dataloader/wsireader.py#L35)

shaneahmed replied:
self.target_means = means
self.target_stds = stds

target_means and target_stds will be class attributes?

shaneahmed reviewed (changes requested):
Overall looks fine to me but I think there is still work which needs to be done. Please can you add the following
- Add attributes in docstring for all the classes. 
- Add command line functions in cli.py
- Update [usage.rst](https://github.com/TIA-Lab/tiatoolbox/blob/feature-stainnorm-ruifrok-reinhard/docs/usage.rst) so the documentation is added to readthedocs.

shaneahmed commented on docs/usage.rst:
```diff
@@ -53,14 +53,32 @@ utils.misc
 .. automodule:: tiatoolbox.utils.exceptions
     :members: FileNotSupported
+
+-----
```
Please can we move Tools before Utils? Just following alphabetical order.

simongraham replied:
Resolved

shaneahmed commented on tiatoolbox/utils/exceptions.py:
```diff
@@ -27,3 +27,11 @@ class FileNotSupported(Exception):
         super().__init__(self.message)
+
+
+class MethodNotSupported(Exception):
```
Can you write a quick test for this code for coverage?

simongraham replied:
Added in[ this commit](https://github.com/TIA-Lab/tiatoolbox/pull/35/commits/a5d3d7f6097a144f92d1e5b1cc183ed1464bfbfd).

shaneahmed commented on tiatoolbox/tools/stainnorm.py:
```diff
@@ -0,0 +1,226 @@
+)
+
+
+class StainNormaliser:
```
Can we not call it RuiforkStainNormaliser?

simongraham replied:
This is a design choice. With the current setup, all that we would need to do in this class is change the `self.extractor`, which is set by the `method` argument. If we do the above, then we will need to define a completely new class.

shaneahmed commented on tiatoolbox/tools/stainnorm.py:
```diff
@@ -0,0 +1,226 @@
+    RuifrokStainExtractor,
+)
+
+
```
Shall we define an ABC?

simongraham replied:
We define ABC for the stain extraction method. Check `tiatoolbox/utils/stainnorm.py`

John-P commented on tiatoolbox/tools/stainnorm.py:
```diff
@@ -0,0 +1,226 @@
+# along with this program; if not, write to the Free Software Foundation,
+# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+#
+# The Original Code is Copyright (C) 2006, Blender Foundation
```
The company name on this line needs to be changed. We are not the Blender Foundation.

simongraham replied:
Fixed in [this commit](https://github.com/TIA-Lab/tiatoolbox/commit/f823ffaac09f575459a9ad590151e1ef7f0bd00e)

shaneahmed commented on tiatoolbox/utils/stainnorm.py:
```diff
@@ -0,0 +1,54 @@
+    """
+
+    @staticmethod
+    def get_stain_matrix(_):
```
Allow the user to input stain matrix.

simongraham replied:
Added CustomNormaliser that allows a user to input their own stain matrix

John-P reviewed (changes requested):
Consider moving stain extraction to a file alongside stainnorm maybe stainextract? Need to think about how we will add more fixed stain matrices for Ruifrok method. Also, are we using Ruifrok to refer to the matrix or the method or both?

simongraham commented:
Following review, changes made:

- Moved stainnorm_utils.py from `utils` directory and renamed to `stainextract.py`
- Restructured the code
- Add helper function `get_normaliser()` to enable smooth usage between methods.
- Add `MethodNotSupported` test
- Update `usage.rst`

shaneahmed commented on tests/test_tiatoolbox.py:
```diff
@@ -289,13 +388,67 @@ def test_background_composite():
+
+    # init class with custom method - test with ruifrok stain matrix
+    stain_matrix = np.array([[0.65, 0.70, 0.29], [0.07, 0.99, 0.11]])
+    norm = get_normaliser("custom", stain_matrix=stain_matrix)
```
I like it!

shaneahmed commented on tiatoolbox/cli.py:
```diff
@@ -269,5 +283,60 @@ def save_tiles(
+@click.option("--target_input", help="input path to the target image")
+@click.option(
+    "--method",
+    help="Stain normlisation method to use. Choose from 'reinhard', 'custom',"
```
For custom method, do you need to pass stain matrix?

simongraham replied:
Yes - will add stain matrix as argument and default to `None` for other methods.

shaneahmed replied:
which argument will be used for passing stain matrix?

simongraham replied:
https://github.com/TIA-Lab/tiatoolbox/blob/feature-stainnorm-ruifrok-reinhard/tiatoolbox/cli.py#L298

shaneahmed replied:
ok great. Please can you add an example how the matrix can be passed as input? I don't think string will work here.

shaneahmed commented on tiatoolbox/tools/stainnorm.py:
```diff
@@ -0,0 +1,282 @@
+    """Stain normalisation base class
+
+    Attributes:
+        stain_matrix_target (ndarray): stain matrix of target
```
How about extractor?

simongraham replied:
Yes, I missed this. Good spot. Pushing and will link the commit.

shaneahmed reviewed (approved):
Thanks Simon. It looks great, we can add Vahadane and Macenko next. Please address the comments I have added and once you get the second review we should be ready to merge.

shaneahmed commented on tiatoolbox/tools/stainnorm.py:
```diff
@@ -0,0 +1,283 @@
+    """Stain normalisation base class
+
+    Attributes:
+        self.extractor: method specific stain extractor
```
It's not added properly. it should be `extractor (type) : method specific ...`

simongraham replied:
okay - this will be an object. Would (obj) be the right type to use here?

shaneahmed replied:
`extractor (CustomExtractor, RuifrokExtractor) : method specific ...`

simongraham replied:
okay - fixed

shaneahmed replied:
In this commit https://github.com/TIA-Lab/tiatoolbox/commit/57aef63606641ae0c3c15f11383380d944c1b4ba

John-P commented on tests/test_tiatoolbox.py:
```diff
@@ -1,13 +1,15 @@
 #!/usr/bin/env python
 
-"""Tests for `tiatoolbox` package."""
+"""pytests for `tiatoolbox` package."""
```
Very minor point. I thought 'Tests' was fine before. Starting the sentence with pytests sounds weird to me. :shrug:

shaneahmed replied:
All we need is to be consistent, whether Tests or pytests

simongraham replied:
This is done for most strings in this file. Shall we change all to Test?

simongraham replied:
I will change all to `Tests` now

simongraham replied:
Fixed [here](https://github.com/TIA-Lab/tiatoolbox/pull/35/commits/0f1081053337cdbaabe8205a779109c9e04d3f81).

John-P commented on tests/test_tiatoolbox.py:
```diff
@@ -101,6 +103,93 @@ def _response_all_wsis(_response_ndpi, _response_svs, tmpdir_factory):
+    return source_file_path
+
+
+@pytest.fixture(scope="session")
+def _response_stainnorm_target(tmpdir_factory):
+    """
+    Sample pytest fixture for target image for stain normalisation
+    Download png image for pytest
+    """
+    target_file_path = tmpdir_factory.mktemp("data").join("target.png")
+    if not os.path.isfile(target_file_path):
+        print("Downloading Target Image for stain normalisation")
+        r = requests.get(
+            "https://warwick.ac.uk/fac/sci/dcs/research/tia/tiatoolbox"
+            "/files/target.png"
+        )
+        with open(target_file_path, "wb") as f:
+            f.write(r.content)
+
+    else:
+        print("Skipping Target Image")
+
+    return target_file_path
```
Since the target image is small, should be not included it as a recourse in the pip package so that it is included with the python code. Just seems a bit silly to download it each time. Especially as it will be used outside of unit tests also.

simongraham replied:
It's up to us I guess. I did it the way you are describing initially and then changed to how it is structured now. Assume that the user will want to supply their own target image? Happy to add an example target image instead within the repo. What do we think?

John-P commented on tiatoolbox/tools/__init__.py:
```diff
@@ -0,0 +1,2 @@
+"""Package defines tools for the toolbox"""
+from tiatoolbox.tools import stainnorm, stainextract
```
I know it's a very small file but missing the GPL header here.

simongraham replied:
This was actually the case with **all** `__init__.py` files. Changed in [this](https://github.com/TIA-Lab/tiatoolbox/pull/35/commits/593da318650b992eee629b982adc3b0f4631a744) commit.

John-P reviewed (changes requested):
Mostly just minor things such as wording and formatting plus one missing GPL header. Otherwise, looks good.
